subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java

```diff
@@ -139,17 +143,20 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository, TaskOutp
         );
     }
 
-    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, TaskExecuter delegate) {
+    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, ListenerManager listenerManager, TaskExecuter delegate) {
         if (startParameter.isTaskOutputCacheEnabled()) {
-            return new SkipCachedTaskExecuter(taskOutputOriginFactory, buildCacheConfiguration, packer, taskOutputsGenerationListener, delegate);
+            TaskCachingHashesListener hashesListener = listenerManager.getBroadcaster(TaskCachingHashesListener.class);
+            TaskCacheKeyCalculator taskCacheKeyCalculator = new TaskCacheKeyCalculator(hashesListener);
+            listenerManager.addListener(new TaskCachingHashesLogger());
```

Comment:
This is another reason for trying to push the cache key calculator out of `SkipCachedTaskExecutor`.  There's getting to be too many things going into it.